### PR TITLE
Implemented async operations

### PR DESCRIPTION
**Pull request by kuvinodms**

### CODE REVIEW

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 1]
```diff
@@ -414,23 +422,77 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
...
+    $RetryIntervalInSeconds = 3
+    if($TimeoutSec)
+    {
+        $MaxTries = [int]($TimeoutSec/$RetryIntervalInSeconds)
+    }
+
+    $timeoutAction = $null
+    if($OperationType -eq $SF_Operations.RegisterApplicationType)
+    {
+        $timeoutAction = {
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Why add an action for this?
You can just put Invoke-ActionWithRetries under try-catch and perform these steps on TimeoutException

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 2]
```diff
@@ -309,6 +311,12 @@ function Register-ServiceFabricApplicationTypeAction
 
     $exceptionRetryEvaluator = {
         param($ex)
+        # If app already created, don't retry
+        if ($ex.GetType().FullName -eq "System.Fabric.FabricElementAlreadyExistsException")
+        {
+            return $false
+        }
+
         $appType = Get-ServiceFabricApplicationTypeAction -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: We don't need to get application type and check its status as it is async call now. that will anyways happen in wait function.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 1 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Yes we can but this is more clear, readable and the word TIMEOUTACTION make more sense to categorize the action that is being performed.
Also this might be helpful for future as well

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 2 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Agreed. Removing it from Register and Unregister both

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1 [hunk 3]
```diff
@@ -206,7 +206,8 @@
                     # It will unregister the existing application's type and version even if its different from the application being created,
                     if ((Get-ServiceFabricApplicationAction | Where-Object {$_.ApplicationTypeVersion -eq $($app.ApplicationTypeVersion) -and $_.ApplicationTypeName -eq $($app.ApplicationTypeName)}).Count -eq 0)
                     {
-                        Unregister-ServiceFabricApplicationTypeAction -ApplicationTypeName $($app.ApplicationTypeName) -ApplicationTypeVersion $($app.ApplicationTypeVersion) -TimeoutSec $UnregisterPackageTimeoutSec
+                        Unregister-ServiceFabricApplicationTypeAction -ApplicationTypeName $($app.ApplicationTypeName) -ApplicationTypeVersion $($app.ApplicationTypeVersion)
+                        Wait-ServiceFabricApplicationTypeRegistrationStatus -ApplicationTypeName $($app.ApplicationTypeName) -ApplicationTypeVersion $($app.ApplicationTypeVersion) -TimeoutSec $UnregisterPackageTimeoutSec
```
[the remark this answers is not in the document]
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-NewServiceFabricApplication.ps1]: I don't wanted that just because both the utility functions has their own retry mechanisms, hence if we move Wait-ServiceFabricApplicationTypeRegistrationStatus inside Unregister-ServiceFabricApplicationTypeAction then for the cases where where Wait-ServiceFabricApplicationTypeRegistrationStatus throw exception then it will be evaluated for Unregister-ServiceFabricApplicationTypeAction as well which can cause issues.

I'm not able to come up with any case where your suggested approach will break so I'm doing changes as suggested by you.

File: Tasks/Common/PowershellHelpers/Helpers.ps1 [hunk 4]
```diff
@@ -24,7 +24,10 @@ function Invoke-ActionWithRetries
         $RetryableExceptions,
 
         [string]
-        $RetryMessage
+        $RetryMessage,
```
[the remark this answers is not in the document]
Comment (kuvinodms) [reply Tasks/Common/PowershellHelpers/Helpers.ps1]: $TimeoutAction looks good and can be used in future also

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 5]
```diff
@@ -309,30 +314,8 @@ function Register-ServiceFabricApplicationTypeAction
...
-        if (!$appType)
-        {
-            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeUnprovisioned)
-            return $true
-        }
-
-        # if app type is provisioned, bail out
-        if ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Available)
+        # If app already created, don't retry
+        if ($ex.GetType().FullName -eq "System.Fabric.FabricElementAlreadyExistsException")
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: This needed to be added to retryable exception

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 6]
```diff
@@ -414,6 +399,63 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
...
+        [string]
+        $OperationType
+    )
+
+    $global:operationId = $SF_Operations.GetApplicationType
+    $getAppTypeAction = { Get-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion }
+    $getAppTypeRetryEvaluator = {
+        param($appType)
+        # if app type does not exist (i.e it got unprovisioned) or if its status has changed to a terminal one, stop wait
+        if ((!$appType) -or (($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -and ($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning)))
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: should you not throw if the appType is null?

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 7]
```diff
@@ -414,6 +399,63 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
...
+        {
+            return $false
+        }
+        else
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeStatus -ArgumentList @($appType.Status, $appType.StatusDetails))
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: what if the application type status is failed

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 8]
```diff
@@ -414,6 +399,63 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
...
+
+    $MaxTries = 1200
+    $RetryIntervalInSeconds = 3
+    if($TimeoutSec)
+    {
+        $MaxTries = [int]($TimeoutSec/$RetryIntervalInSeconds)
+    }
+
+    $timeoutAction = $null
+    if($OperationType -eq $SF_Operations.RegisterApplicationType)
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Dont  try to have if/else based on the operation in the common method. Instead have the logic separately in the caller

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 9]
```diff
@@ -414,6 +399,63 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
...
+    {
+        $timeoutAction = {
+            Unregister-ServiceFabricApplicationTypeAction -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
+            Wait-ServiceFabricApplicationTypeRegistrationStatus -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion -TimeoutSec $TimeoutSec
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: wont unregister fail if the registration is in progress?

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 5 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Added

Comment (sachinma) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: This still wont cause the retry for the transient exceptions.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 10]
```diff
@@ -309,50 +314,34 @@ function Register-ServiceFabricApplicationTypeAction
...
     }
     catch
     {
         # print cluster health status if registering failed
         Trace-ServiceFabricClusterHealth
         throw
     }
+
+    $timeoutAction = {
+        Unregister-ServiceFabricApplicationTypeAction -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion -TimeoutSec $TimeoutSec
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Will the unregister work when the registration is in progress?

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 11]
```diff
@@ -414,73 +403,118 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
...
 
         [int]
-        $TimeoutSec
+        $TimeoutSec,
+
+        [string]
+        $TimeoutAction,
+
+        [string]
+        $OperationType
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Same as last time - i dont see much value if you need to pass the caller information to make decisions in the common method

File: Tasks/Common/PowershellHelpers/Helpers.ps1 [hunk 12]
```diff
@@ -24,7 +24,10 @@ function Invoke-ActionWithRetries
...
+
+        [scriptblock]
+        $TimeoutAction
```
Comment (bishal-pdMSFT) [inline Tasks/Common/PowershellHelpers/Helpers.ps1]: If you want to keep an action which is run when an exception happens, then it should be generic enough i.e it should handle exceptions other than timeout.
I still believe this is not the right approach.

File: Tasks/Common/PowershellHelpers/Helpers.ps1 [hunk 13]
```diff
@@ -76,7 +79,14 @@ function Invoke-ActionWithRetries
             }
             else
             {
-                throw (Get-VstsLocString -Key ActionTimedOut)
+                if($TimeoutAction)
+                {
+                    $TimeoutAction.Invoke()
```
Comment (bishal-pdMSFT) [inline Tasks/Common/PowershellHelpers/Helpers.ps1]: Should you throw here? or take that as input?

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 10 above)
Comment (bishal-pdMSFT) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: +1
This should be a best-effor. suppress any error

Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: re-throw original exception

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 14]
```diff
@@ -414,73 +403,118 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
...
 
-    $global:operationId = $SF_Operations.UnregisterApplicationType
+    $global:operationId = $SF_Operations.GetApplicationType
+    $getAppTypeAction = { Get-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion }
+    $getAppTypeRetryEvaluator = {
+        param($appType)
 
-    $unregisterAction = {
-        Unregister-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion -Force -TimeoutSec $TimeoutSec
-        if (!$?)
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: this if block should be kept in the Unregister-ServiceFabricApplicationTypeAction's $registerAction delegate

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 15]
```diff
@@ -414,73 +403,118 @@ function Wait-ServiceFabricApplicationTypeTerminalStatus
         -RetryMessage (Get-VstsLocString -Key SFSDK_RetryingGetApplicationType)
 }
 
-function Unregister-ServiceFabricApplicationTypeAction
+function Wait-ServiceFabricApplicationTypeRegistrationStatus
```
Comment (bishal-pdMSFT) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Why not have two wait functions - on which waits for register to complete and another which waits for unregister to complete

File: Tasks/Common/PowershellHelpers/Helpers.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (kuvinodms) [reply Tasks/Common/PowershellHelpers/Helpers.ps1]: Cool, lets remove it.

File: Tasks/Common/PowershellHelpers/Helpers.ps1 (hunk 12 above)
Comment (kuvinodms) [reply Tasks/Common/PowershellHelpers/Helpers.ps1]: Cool, lets remove it.

File: Tasks/Common/PowershellHelpers/Helpers.ps1 (hunk 13 above)
Comment (kuvinodms) [reply Tasks/Common/PowershellHelpers/Helpers.ps1]: Removing this logic

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 10 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Suppressing now.

Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Yes

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 15 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Creating 2 functions as suggested.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 11 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Okay

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 5 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Default return is $true for this function, hence it will always retry for transient and timeout exceptions

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 14 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Its not being touched as a part of this PR

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 16]
```diff
@@ -427,60 +421,134 @@ function Unregister-ServiceFabricApplicationTypeAction
...
+
+        [int]
+        $TimeoutSec
+    )
+
+    $global:operationId = $SF_Operations.UnregisterApplicationType
+
+    $unregisterAction = {
+        Unregister-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion -Async -Force
+        if (!$?)
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: why is this required? Wont unregister-service.. throw?

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 17]
```diff
@@ -427,60 +421,134 @@ function Unregister-ServiceFabricApplicationTypeAction
...
+
+        [string]
+        $ApplicationTypeVersion,
+
+        [int]
+        $TimeoutSec
+    )
+
+    $global:operationId = $SF_Operations.GetApplicationType
+    $getAppTypeAction = { Get-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion }
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Use Action here i.e. Get-ServiceFabricApplicationTypeAction

Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Can this throw if the ApplicationType doesn't exist or does it return null

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 18]
```diff
@@ -427,60 +421,134 @@ function Unregister-ServiceFabricApplicationTypeAction
...
-    $global:operationId = $SF_Operations.UnregisterApplicationType
+    $global:operationId = $SF_Operations.GetApplicationType
+    $getAppTypeAction = { Get-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion }
+    $getAppTypeRetryEvaluator = {
+        param($appType)
 
-    $unregisterAction = {
-        Unregister-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion -Force -TimeoutSec $TimeoutSec
-        if (!$?)
+        # If provisioning not started, retry register
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Since this is wait, you should throw if appType is null

Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: As discussed, this is correct behavior.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 17 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Sure

Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: This will return null

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 16 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: As discussed, this will ensure if above command ran successfully. However this check is present from long back and we don't want to modify its behavior.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 19]
```diff
@@ -427,60 +421,134 @@ function Unregister-ServiceFabricApplicationTypeAction
...
         }
-    }
-
-    $exceptionRetryEvaluator = {
-        param($ex)
-        $appType = Get-ServiceFabricApplicationTypeAction -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
-        # If app type already unprovisioned, don't retry
-        if (!$appType)
+        # if app type is provisioned, don't retry
+        elseif($appType -and ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Available))
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: You don't need to put $appType check here since it would not come here appType is null

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 20]
```diff
@@ -427,60 +421,134 @@ function Unregister-ServiceFabricApplicationTypeAction
...
         {
-            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeUnprovisioningStarted)
-            $appType = Wait-ServiceFabricApplicationTypeTerminalStatus -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
+            return $true
         }
+        # if app type exist and if its status has changed to a terminal one, throw
+        elseif($appType -and (($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -and ($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning)))
+        {
+            throw (Get-VstsLocString -Key SFSDK_RegisterAppTypeFailedWithStatus -ArgumentList @($appType.Status, $appType.StatusDetails))
+        }
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: What if none of these conditions are met. What is the default return value?

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 [hunk 21]
```diff
@@ -427,60 +421,134 @@ function Unregister-ServiceFabricApplicationTypeAction
...
+        # if app type exist and if its status has not changed to a terminal one, do retry
+        elseif($appType -and (($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -or ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning)))
         {
             return $true
         }
+        # if app type exist and if its status has changed to a terminal one, throw
+        elseif($appType -and (($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -and ($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning)))
+        {
+            throw (Get-VstsLocString -Key SFSDK_RegisterAppTypeFailedWithStatus -ArgumentList @($appType.Status, $appType.StatusDetails))
+        }
```
Comment (sachinma) [inline Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: same as before

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 19 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Correct

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 20 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: If you look at semantics of last 2 ELSEIF statements than you will realize that last ELSEIF is nothing but an ELSE statement. Hence this will always work as default return.

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 21 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: same as before

File: Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1 (hunk 20 above)
Comment (kuvinodms) [reply Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1]: Have done it intentionally to enhance readability of these conditions. Also, it doesn't harm in any way.
